### PR TITLE
Update out-of-bounds-access to be non-interactive

### PR DESCRIPTION
**Pull request by jonmattgray**
Previously the `out-of-bounds-access` example was interactive and awaited user input on the command line. This changes the example to be non-interactive and take an even number of command line arguments as username and password pairs. 

e.g. `$./out-of-bounds-access root------------123 bla root 123` will successfully login the second time as the password is changed to `123` by the first `root------------123` attempt.

image

### CODE REVIEW

File: src/out-of-bounds-access.c [hunk 1]
```diff
@@ -3,56 +3,64 @@
...
+int main(int argc, char *argv[])
 {
+    if ((argc - 1) % 2 != 0) // argc counts program name as first argument
+    {
+        printf("Must be supplied an even number of arguments\n");
+        return 1;
+    }
+
     char username[] = "root";
     char password[] = "password";
```
Comment (jonmattgray) [inline src/out-of-bounds-access.c]: I tried making these global so they don't have to be passed in for every login attempt, but then the memory allocation isn't adjacent with `buffer[12]`, which is needed for the out of bounds write to work.

File: src/out-of-bounds-access.c [hunk 2]
```diff
@@ -3,56 +3,64 @@
...
 }
 
-int main(void)
+int main(int argc, char *argv[])
 {
+    if ((argc - 1) % 2 != 0) // argc counts program name as first argument
+    {
+        printf("Must be supplied an even number of arguments\n");
+        return 1;
+    }
```
Comment (mattdean-digicatapult) [inline src/out-of-bounds-access.c:45]: nice guarding!

File: src/out-of-bounds-access.c (hunk 1 above)
Comment (mattdean-digicatapult) [reply src/out-of-bounds-access.c]: maybe you can make the pointers global but `malloc` them to ensure allocation order?

File: src/out-of-bounds-access.c [hunk 3]
```diff
@@ -3,56 +3,64 @@
...
+    char buffer[12];
 
     // check memory allocation is adjacent
-    assert((void *)&password[0] == (void *)&input[sizeof(input)]);
-
-    printf("Password is: %s\n\n", password);
+    assert((void *)&password[0] == (void *)&buffer[sizeof(buffer)]);
 
-    while (1)
+    for (int i = 1; i < argc; i += 2)
```
Comment (mattdean-digicatapult) [inline src/out-of-bounds-access.c:59]: I'd be tempted to move the "copy" from `argv[i]` to `buffer` here. That makes the bug a bit more believable. If you were to have two buffers we could probably make the pointers global (see comment in thread) and that would make `runAttempt` take no args.

File: src/out-of-bounds-access.c (hunk 1 above)
Comment (jonmattgray) [reply src/out-of-bounds-access.c]: done

File: src/out-of-bounds-access.c (hunk 3 above)
Comment (jonmattgray) [reply src/out-of-bounds-access.c:59]: done

File: src/out-of-bounds-access.c [hunk 4]
```diff
@@ -3,56 +3,68 @@
...
-        return 0;
+        printf("%s\n", "Login failed");
+        return 1;
     }
-    else
+
+    if (strcmp(passwordBuffer, password))
     {
-        printf("Correct input\n");
+        printf("%s\n", "Login failed");
```
Comment (mattdean-digicatapult) [inline src/out-of-bounds-access.c:31]: Should this be `Login succeeded`?

File: src/out-of-bounds-access.c [hunk 5]
```diff
@@ -3,56 +3,68 @@
...
 
-    while (1)
+    for (int i = 1; i < argc; i += 2)
     {
-        runAttempt(input, username, password);
+        char *usernameAttempt = argv[i];
+        char *passwordAttempt = argv[i + 1];
+        strcpy(usernameBuffer, usernameAttempt);
+        strcpy(passwordBuffer, passwordAttempt);
+        loginAttempt();
```
Comment (mattdean-digicatapult) [inline src/out-of-bounds-access.c:65]: this looks really nice now

**Review by mattdean-digicatapult — APPROVED** [review]

File: src/out-of-bounds-access.c (hunk 5 above)
Comment (mattdean-digicatapult) [reply src/out-of-bounds-access.c:65]: sorry misread the confusing diff

**Review by n3op2 — APPROVED** [review]
(party)

**Pull request merged**